Issue 4: Enforce authentication with JWT

twistedstream opened the pull request:
Lua implementation is in and all integration tests are passing!  Still need to touch up the README to provide install instructions, which should be done shortly.  In the meantime, PR is here for review.

Please merge PR #14 first.

Closes #4

twistedstream committed:
Pull down Lua dependencies

and copy them to the proxy Docker container

twistedstream committed:
Use resty.jwt to verify the JWT token

Pass the secret into the module function from nginx.conf

twistedstream committed:
Write the X-Auth-UserId header

with valid JWT token

twistedstream committed:
Rename Usage section to Installation in README

and move it to the near top so people can quickly see how to use the module

twistedstream committed:
Simplify Contributing section

and add Node.js prerequisite

twistedstream committed:
Tweak JWT secret

so it doesn't include a single quote

twistedstream committed:
jwt.auth() now uses an env variable for secret

instead of passing it in as a param.
This makes for easier configuration in the nginx.conf file, plus its arguably a better practice to store the secret in an env variable.

twistedstream commented:
README has been updated, plus I moved to sourcing the JWT secret from an environment variable.

woloski commented on nginx-jwt.lua:
```diff
@@ -0,0 +1,44 @@
+                ngx.log(ngx.INFO, "JWT: " .. cjson.encode(jwt_obj))
+
+                -- write the X-Auth-UserId header 
+                ngx.header["X-Auth-UserId"] = jwt_obj.payload.sub
```
We should support specifying which props you want to passthrogh as headers (through config)

twistedstream replied:
Agreed.  And the only reason I even send back the X-Auth-UserId header in the first place is so that the [integration test](https://github.com/twistedstream/nginx-jwt/blob/issue-4/test/test_integration.js#L37) can assert that the JWT was actually verified.

twistedstream committed:
Merge branch 'master' into issue-4

twistedstream committed:
Support URL-base-64 encoded secrets

using basexx (https://github.com/aiq/basexx) Lua library to do the base64 decoding.

twistedstream commented:
The module now supports URL-safe base-64 encoded JWT secrets, enabled via the `JWT_SECRET_IS_BASE64_ENCODED` environment variable

twistedstream committed:
Merge branch 'master' into issue-4

twistedstream committed:
Integration tests handle normal and base64 secrets

twistedstream committed:
Enhance README

primarily to describe the project as a Lua script instead of an Nginx module

twistedstream commented:
@woloski Added integration tests that test for both normal and Base64-encoded secrets
